hbase-client/src/main/java/org/apache/hadoop/hbase/security/provider/DefaultProviderSelector.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.security.provider;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.security.User;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.Private
+public class DefaultProviderSelector implements AuthenticationProviderSelector {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultProviderSelector.class);
+  private final ReadWriteLock LOCK = new ReentrantReadWriteLock();
+
+  Configuration conf;
+  SimpleSaslClientAuthenticationProvider simpleAuth = null;
+  GssSaslClientAuthenticationProvider krbAuth = null;
+  DigestSaslClientAuthenticationProvider digestAuth = null;
+
+  @Override
+  public void configure(Configuration conf, Map<Byte,SaslClientAuthenticationProvider> providers) {
+    try {
+      LOCK.writeLock().lock();
+      if (this.conf != null) {
+        throw new IllegalStateException("configure() should only be called once");
+      }
+      this.conf = Objects.requireNonNull(conf);
+    } finally {
+      LOCK.writeLock().unlock();
+    }
+    
+    for (SaslClientAuthenticationProvider provider : Objects.requireNonNull(providers).values()) {
+      if (provider instanceof SimpleSaslClientAuthenticationProvider) {
```

Comment:
Is it possible to use some design pattern like visitor pattern to avoid the instance of here? Not a blocker one anyway, just a bit ugly, could have a try.